src/cmd/metadata/mod.rs

```diff
@@ -0,0 +1,367 @@
+            test_manifest.add_package_value("license", "Apache-2.0".into())?;
+            test_manifest
+                .add_user_metadata_value("some-user-provided-field", "and-its-value".into())?;
+            test_manifest.add_user_metadata_value(
+                "more-user-provided-fields",
+                vec!["and", "their", "values"].into(),
+            )?;
+            test_manifest.write()?;
+
+            let crate_metadata = CrateMetadata::collect(&test_manifest.manifest_path)?;
+            let metadata_file =
+                cmd::metadata::execute(test_manifest.manifest_path, None, UnstableFlags::default())
+                    .expect("generate metadata failed");
+            let metadata_json: Map<String, Value> =
+                serde_json::from_slice(&fs::read(&metadata_file)?)?;
+
+            assert!(
+                metadata_file.exists(),
+                format!("Missing metadata file '{}'", metadata_file.display())
+            );
+
+            let source = metadata_json.get("source").expect("source not found");
+            let hash = source.get("hash").expect("source.hash not found");
+            let language = source.get("language").expect("source.language not found");
+            let compiler = source.get("compiler").expect("source.compiler not found");
+
+            let contract = metadata_json.get("contract").expect("contract not found");
+            let name = contract.get("name").expect("contract.name not found");
+            let version = contract.get("version").expect("contract.version not found");
+            let authors = contract
+                .get("authors")
+                .expect("contract.authors not found")
+                .as_array()
+                .expect("contract.authors is an array")
+                .iter()
+                .map(|author| author.as_str().expect("author is a string"))
+                .collect::<Vec<_>>();
+            let description = contract
+                .get("description")
+                .expect("contract.description not found");
+            let documentation = contract
+                .get("documentation")
+                .expect("contract.documentation not found");
+            let repository = contract
+                .get("repository")
+                .expect("contract.repository not found");
+            let homepage = contract
+                .get("homepage")
+                .expect("contract.homepage not found");
+            let license = contract.get("license").expect("contract.license not found");
+
+            let user = metadata_json.get("user").expect("user section not found");
+
+            // calculate wasm hash
+            let wasm = fs::read(&crate_metadata.dest_wasm)?;
+            let mut output = [0u8; 32];
+            let mut blake2 = blake2::VarBlake2b::new_keyed(&[], 32);
+            blake2.update(wasm);
+            blake2.finalize_variable(|result| output.copy_from_slice(result));
```

Comment:
Same as above.

Reply:
Also not sure why you test a functionality by some other functionality with the exact same implementation. Normally when testing hashes I do so with some concrete hashes that I compute once and then take for granted.

Reply:
It is essentially just testing that we are using that same hash function to generate the hash. Also a hardcoded hash could cause this test to break a lot with different versions of `ink!` and `rustc`.